### PR TITLE
[SPARK-18986][Core] ExternalAppendOnlyMap shouldn't fail when forced to spill before calling its iterator

### PR DESCRIPTION
**Pull request by viirya**
## What changes were proposed in this pull request?

`ExternalAppendOnlyMap.forceSpill` now uses an assert to check if an iterator is not null in the map. However, the assertion is only true after the map is asked for iterator. Before it, if another memory consumer asks more memory than currently available, `ExternalAppendOnlyMap.forceSpill` is also be called too. In this case, we will see failure like this:

    [info]   java.lang.AssertionError: assertion failed
    [info]   at scala.Predef$.assert(Predef.scala:156)
    [info]   at org.apache.spark.util.collection.ExternalAppendOnlyMap.forceSpill(ExternalAppendOnlyMap.scala:196)
    [info]   at org.apache.spark.util.collection.Spillable.spill(Spillable.scala:111)
    [info]   at org.apache.spark.util.collection.ExternalAppendOnlyMapSuite$$anonfun$13.apply$mcV$sp(ExternalAppendOnlyMapSuite.scala:294)

This fixing is motivated by http://apache-spark-developers-list.1001551.n3.nabble.com/java-lang-AssertionError-assertion-failed-tc20277.html.

## How was this patch tested?

Jenkins tests.

Please review http://spark.apache.org/contributing.html before opening a pull request.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/util/collection/ExternalAppendOnlyMap.scala [hunk 1]
```diff
@@ -192,12 +193,16 @@ class ExternalAppendOnlyMap[K, V, C](
...
-      currentMap = null
+    if (isReadingIterator) {
+      assert(readingIterator != null)
+      val isSpilled = readingIterator.spill()
+      if (isSpilled) {
+        currentMap = null
+      }
+      isSpilled
+    } else {
+      false
```
Comment (viirya) [inline core/src/main/scala/org/apache/spark/util/collection/ExternalAppendOnlyMap.scala]: I choose to simply return false now. Another option is to actually spill the in-memory map.

**Comment by viirya** [comment]
retest this please.

**Comment by viirya** [comment]
cc @JoshRosen Can you take a look? Thanks.

**Comment by viirya** [comment]
cc @rxin @zsxwing too

**Comment by viirya** [comment]
ping @rxin @zsxwing @JoshRosen @davies

File: core/src/main/scala/org/apache/spark/util/collection/ExternalAppendOnlyMap.scala [hunk 2]
```diff
@@ -115,6 +115,7 @@ class ExternalAppendOnlyMap[K, V, C](
   private val keyComparator = new HashComparator[K]
   private val ser = serializer.newInstance()
 
+  @volatile private var isReadingIterator: Boolean = false
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/util/collection/ExternalAppendOnlyMap.scala]: I'm a little confused. Isn't this having the same effect as just removing the assert, since you're setting this to `true` right after instantiating `readingIterator`?

Comment (viirya) [inline core/src/main/scala/org/apache/spark/util/collection/ExternalAppendOnlyMap.scala]: Yeah, alternatively we can remove the assert and check if `readingIterator` is null or not. I just want to keep the original behavior.

**Comment by viirya** [comment]
retest this please.

**Comment by viirya** [comment]
@vanzin Is this looking good for you now? Thanks.

File: core/src/main/scala/org/apache/spark/util/collection/ExternalAppendOnlyMap.scala (hunk 1 above)
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/util/collection/ExternalAppendOnlyMap.scala]: This fixes the problem reported in the bug, but is there a reason not to just instantiate `readingIterator` if it hasn't been set yet? Feels like it would be the right thing to do.

Comment (viirya) [inline core/src/main/scala/org/apache/spark/util/collection/ExternalAppendOnlyMap.scala]: As I commented before:

> I choose to simply return false now. Another option is to actually spill the in-memory map.

If `readingIterator` is not set, once `forceSpill` is called we can spill the content of `currentMap` into disk like `spill` does now.

We can't just instantiate `readingIterator` and spill it. Because `readingIterator` is only created for iterating over the entries of the in-memory map, when we want to read the content of this `ExternalAppendOnlyMap`. If `forceSpill` is called before we want to read the content, and we instantiate `readingIterator` and spill it, the in-memory map will be released and we can't insert data further.

Comment (vanzin) [inline core/src/main/scala/org/apache/spark/util/collection/ExternalAppendOnlyMap.scala]: Sorry, you're right, so let me rephrase: is there any reason not to try to spill the current map if the iterator hasn't been instantiated?

Comment (viirya) [inline core/src/main/scala/org/apache/spark/util/collection/ExternalAppendOnlyMap.scala]: The reason I chose to simple return false is, if the in-memory map is not yet spilled to disk during inserting data, it may not need to be spilled in fact as it should be a small spill.

But I think it is arguable that when `forceSpill` is called, we should spill the collection anyway. I will do the change.

Comment (viirya) [inline core/src/main/scala/org/apache/spark/util/collection/ExternalAppendOnlyMap.scala]: BTW, `ExternalSorter` has the same behavior as the current change: its `forceSpill` does nothing before its iterator is called.

**Comment by vanzin** [comment]
LGTM, I'll leave it open a bit in case someone else has comments.

**Comment by viirya** [comment]
@vanzin ok, thanks for review.

**Comment by samkum** [comment]
I have tested this, but I found a very strange observation. GC frequency has increased many folds...and majority of the time is spend in GC.

**Comment by viirya** [comment]
@samkum Thanks for testing this. I think it is because every time `forceSpill` is called now, it will spill the map anyway. I will add a check to only spill the map if the map is not empty.

**Comment by viirya** [comment]
retest this please.

**Comment by samkum** [comment]
The fix is better but still produces visible GC issues. After removing the fix, GC is no longer a problem.

gc

**Comment by viirya** [comment]
@samkum I've expected this and your testing inspired me to change how the memory consumers are spilled in order. I've submitted a PR #16603 for this.

Simply said, even I added a check and only spill the map if it has elements. Currently, the map will be chose to spill even it has few elements added since last time it spilled.

Because of that, I will hope #16603 can be reviewed and merged first before this. So we won't see the GC issue.

Can you test #16603 and see if it helps GC issue of this PR? Thanks.

File: core/src/main/scala/org/apache/spark/util/collection/ExternalAppendOnlyMap.scala (hunk 1 above)
Comment (viirya) [inline core/src/main/scala/org/apache/spark/util/collection/ExternalAppendOnlyMap.scala]: @vanzin As @samkum tested, spilling the map could cause GC pressure. I think that is because the map is frequently spilled.

Because the map will be checked and spilled during `insertAll`, I think it is ok not to spill it in `forceSpill` in order to avoid frequent spilling.

What do you think?

Comment (vanzin) [inline core/src/main/scala/org/apache/spark/util/collection/ExternalAppendOnlyMap.scala]: I'm not sure I understand how freeing memory would cause GC pressure, unless you're not actually freeing memory.

@samkum did you try this patch without the "spill map on every call" behavior (basically returning "false" instead of spilling the in-memory map when the iterator doesn't exist)?

If that case also results in increased gc, maybe your app just needs more memory. That would be the case where the assert would trigger before.

Comment (viirya) [inline core/src/main/scala/org/apache/spark/util/collection/ExternalAppendOnlyMap.scala]: @samkum Do you have any comment on @vanzin's question?

**Comment by vanzin** [comment]
@samkum  can you reply to the questions above? This PR is basically blocked on your feedback that it makes things worse.

> @samkum did you try this patch without the "spill map on every call" behavior (basically returning "false" instead of spilling the in-memory map when the iterator doesn't exist)?
>
> If that case also results in increased gc, maybe your app just needs more memory. That would be the case where the assert would trigger before.

**Comment by samkum** [comment]
Hi,

No, I havent test this with out the ExternalAppendOnlyMap fix if this is
what you are asking. I tested it with changes to both the classes:
TaskMemoryManager and ExternalAppendOnlyMap.

-Sameer.

**Comment by vanzin** [comment]
No the question is whether you tested without @viirya commit `b1ef9ec` (the last one that forces spills of in-memory maps), or just the very last version of the patch.

**Comment by samkum** [comment]
Nope, I didn't tested it in isolation.

-Sameer.

**Comment by vanzin** [comment]
> Nope, I didn't tested it in isolation.

Could you do that? To make sure that it's really caused by this change?

**Comment by samkum** [comment]
ok..I will get back to you in next couple of days.

-Sameer.

**Comment by samkum** [comment]
@mridulm <https://github.com/mridulm> - need one clarification: if I test
it with out ut @viirya <https://github.com/viirya> commit b1ef9ec,wouldnt
the commit fail at the assertion.( assert(readingIterator != null). Would
the taskManager change take care of this.

**Comment by vanzin** [comment]
> wouldnt the commit fail at the assertion

No. We want to know if testing without that commit shows the same GC behavior you saw with that commit. There's no assert in the code when you just remove that commit.

**Comment by viirya** [comment]
@samkum Any update?

**Comment by samkum** [comment]
Hi liang,

I got pulled into another project. Will need few more days to catch up on
bandwidth.

Just for confirmation, I will testing with changes just to TaskManager
class.

-Sameer.

**Comment by vanzin** [comment]
> I will testing with changes just to TaskManager class.

No, we're asking you to test without the last two commits in this PR. Alternatively, if possible, you could post the code somewhere so that we don't have to depend on you to try things out.

**Comment by samkum** [comment]
Please share the commit id...i think that will make it clear. I can't share the code as its propriety to organization,. but i am just reducing a map phase, reduce by key and reducebykeyandwindow.

**Comment by vanzin** [comment]
@samkum the code is in this PR! Just revert the last two commits.

**Comment by samkum** [comment]
Hi Marcelo,

Are you referring to following commit.

5645533(
https://github.com/apache/spark/pull/16387/commits/5645533ede8bb37aa316afbc0d003d042bf5d52d
)

-Sameer.

**Comment by vanzin** [comment]
Go to the top of this page. Click on the "commits" tab. Look at the last two on the list. Remove those ("git reset" or "git revert" or whatever you prefer). Build, test.

**Comment by vanzin** [comment]
@samkum if you can't get back to us with the information we asked for, I'll just assume that the problem is on your side and push this change. It's been sitting here already reviewed for long enough.

**Comment by samkum** [comment]
I can't commit on the timelines. I don't want to block things at my end.

-Sameer.

**Comment by vanzin** [comment]
Alright. I'm merging this to master, we can always fix things later if there's an issue.

Sorry for the delay @viirya

**Comment by viirya** [comment]
Thanks! @vanzin 
Yes we can always come back to fix things later if there is an issue.
